pkg/streamingpromql/functions.go

```diff
@@ -251,6 +252,42 @@ func ClampMinMaxFunctionOperatorFactory(functionName string, isMin bool) Instant
 	}
 }
 
+func RoundFunctionOperatorFactory() InstantVectorFunctionOperatorFactory {
+	return func(args []types.Operator, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, _ *annotations.Annotations, expressionPosition posrange.PositionRange, timeRange types.QueryTimeRange) (types.InstantVectorOperator, error) {
+		if len(args) != 1 && len(args) != 2 {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected exactly 1 or 2 argument for round, got %v", len(args))
+		}
+
+		inner, ok := args[0].(types.InstantVectorOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected an instant vector for 1st argument for round, got %T", args[0])
+		}
+
+		var toNearest types.ScalarOperator
+		if len(args) == 2 {
+			toNearest, ok = args[1].(types.ScalarOperator)
+			if !ok {
+				// Should be caught by the PromQL parser, but we check here for safety.
+				return nil, fmt.Errorf("expected a scalar for 2nd argument for round, got %T", args[1])
+			}
+		} else {
+			toNearest = operators.NewScalarConstant(float64(1), timeRange, memoryConsumptionTracker, expressionPosition)
+		}
+
+		f := functions.FunctionOverInstantVector{
+			SeriesDataFunc: functions.Round,
+			// TODO(jhesketh): With the currently vendored prometheus, round does not consistently drop the __name__ label
```

Comment:
Does not consistently drop the `__name__` label, or doesn't drop it at all?

Is there an issue link for this?

Reply:
This is tricky. Against main, round does not return any __name__ (as evidence by the tests here https://github.com/prometheus/prometheus/pull/15176).

Against what we have vendored it does have the __name__. So the bug has seemingly since been fixed. I have not found an exact issue but assume it is related to the more recent delayed label name dropping work fixing something incidentally.

Also worth noting is that in the vendored version it has `DropName: true,`, so I'm not sure if that isn't being honoured somewhere (I haven't tracked down where). 

The only upstream tests of round are in the aggregations.test file, such as:
```
eval instant at 50m round(0.005 * http_requests{group="production",job="api-server"})
	{group="production", instance="0", job="api-server"} 1
	{group="production", instance="1", job="api-server"} 1
```
Which due to the multiplication has the resulting label dropped.

So it wouldn't surprise me if it went unnoticed upstream and also fixed as part of the label work (also unnoticed).

Reply:
Right, thanks for the explanation.

> Also worth noting is that in the vendored version it has `DropName: true,`, so I'm not sure if that isn't being honoured somewhere (I haven't tracked down where).

`DropName` is only used when the experimental new delayed `__name__` dropping is enabled, and this is disabled by default.


I would rephrase the comment slightly to make it clearer, something like this perhaps:

```suggestion
// TODO(jhesketh): With the version of Prometheus vendored at the time of writing, round does not drop the __name__ label, and this is verified by our tests.
//                 We match this behaviour for consistency, but this behaviour has changed in Prometheus 3.0, so we'll need to match that once prometheus 3.0 is vendored in.
```